src/tbb/arena.h

```diff
@@ -179,6 +179,82 @@ class atomic_flag {
     }
 };
 
+#if __TBB_PREVIEW_PARALLEL_PHASE
+class thread_leave_manager {
+    static const std::uintptr_t DELAYED_LEAVE       = 0;
+    static const std::uintptr_t FAST_LEAVE          = 1;
+    static const std::uintptr_t ONE_TIME_FAST_LEAVE = 1 << 1;
+    static const std::uintptr_t PARALLEL_PHASE      = 1 << 2;
+
+    std::atomic<std::uintptr_t> my_state{static_cast<std::uintptr_t>(-1)};
+public:
+    // This method is not thread-safe!
+    // Required to be called after construction to set initial state of the state machine.
+    void set_initial_state(tbb::task_arena::leave_policy lp) {
+        if (lp == tbb::task_arena::leave_policy::automatic) {
+            std::uintptr_t platform_policy = governor::hybrid_cpu() ? FAST_LEAVE : DELAYED_LEAVE;
+            my_state.store(platform_policy, std::memory_order_relaxed);
+        } else {
+            __TBB_ASSERT(lp == tbb::task_arena::leave_policy::fast,
+                         "Was the new value introduced for leave policy?");
+            my_state.store(FAST_LEAVE, std::memory_order_relaxed);
+        }
+    }
+
+    void reset_if_needed() {
+        std::uintptr_t curr = ONE_TIME_FAST_LEAVE;
+        if (my_state.load(std::memory_order_relaxed) == curr) {
+            // Potentially can override decision of the parallel block from future epoch
+            // but it is not a problem because it does not violate the correctness
+            my_state.compare_exchange_strong(curr, DELAYED_LEAVE);
+        }
+    }
```

Comment:
In reset, you only need to drop the one-time flag, correct? Since now it does not intersect with other states, it should be enough to just `fetch_and(~ONE_TIME_FAST_LEAVE)`.

Reply:
I think the desire is to drop the one-time flag only if a new work is submitted in the empty arena. So, the flag cannot be blindly dropped.

Reply:
Well, at first glance it seems like @akukanov suggestion shouldn't contradict the desire. Since `unregister_parallel_phase` sets this bit only on the last parallel phase completion, when new work submission happens the state will be either still `ONE_TIME_FAST_LEAVE`, or already include some amount of parallel phase references, which means that this bit will be already dropped by some of the first parallel phases.

Reply:
Right, if the arena is already not empty, then the flag should have been dropped already, right? And then `fetch_and` would just do nothing. We can still leave the load and the check if it gives a proven performance advantage, but certainly `compare_exchange_strong` is not needed.

Reply:
Replaced CAS by `fetch_and`.